File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1 for _ in range(n)]
        dp[-1] = 0
        def recSearch(i: int) -> int:
            nonlocal dp, n
            
            if dp[i] == -1:  # never calculated
                maxJump = nums[i]
                res = n + 1
                for j in range(min(maxJump, n-1-i), 0, -1):
                    res = min(res, 1 + recSearch(i+j))
                dp[i] = res
            return dp[i]

        return recSearch(0)

File: 0045-jump-game-ii/0045-jump-game-ii.rs
use std::cmp;


impl Solution {
    pub fn recSearch(i: usize , nums: &Vec<i32>, n: &i32, dp: &mut Vec<i32>) -> i32 {
        if (dp[i] == -1) {  // never calculated
            let maxJump: i32 = nums[i];
            let mut res: i32 = *n + 1;
            for j in 1..(maxJump+1) as usize {
                if ((i + j) < (*n as usize)) {
                    res = cmp::min(res, 1 + Solution::recSearch(i+j, nums, n, dp));
                }
            }
            dp[i] = res;
        }
        return dp[i];
    }
    pub fn jump(nums: Vec<i32>) -> i32 {
        let n: i32 = nums.len() as i32;
        let mut dp: Vec<i32> = Vec::new();
        for i in 0..n {
            dp.push(-1);
        }
        dp[(n-1) as usize] = 0;

        return Solution::recSearch(0 as usize, &nums, &n, &mut dp);
    }
}
